### DISCUSSION TITLE
client unexpectedly closed TCP connection with status check

### DISCUSSION THREAD

**Discussion by unclewizard** [celery/celery#7028] [Q&A] [answered]
## Checklist

```
software -> celery:4.1.0 (latentcall) kombu:4.1.0 py:2.7.12
            billiard:3.5.0.3 py-amqp:2.2.2
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:disabled
```

## Steps to reproduce

Follow the "first steps with django" documentation to create a barebones celery app: http://docs.celeryproject.org/en/latest/django/first-steps-with-django.html

For a broker, use RabbitMQ 3.6.11. I have also tested with RabbitMQ 3.6.12. I set my broker in settings.py as follows:

```
BROKER_URL = 'amqp://guest:guest@localhost:5672/'
```

Finally, run ``celery worker -A appname status`` 

## Expected behavior

No warnings in the broker logs

## Actual behavior

The rabbitmq server prints the following output (depending on configuration it may show up in the log file. For me it was in /var/log/rabbitmq/rabbit@vagrant-ubuntu-trusty-64.log)

```
=WARNING REPORT==== 31-Oct-2017::23:56:19 ===
closing AMQP connection <0.663.0> (127.0.0.1:35985 -> 127.0.0.1:5672, vhost: '/', user: 'guest'):
client unexpectedly closed TCP connection
```

The status check reports "OK" and the worker is consuming tasks normally, so this doesn't appear to be a connectivity or network issue.

This wouldn't be so bad, but it also seems to trigger a memory leak at the broker. If you run this 1000s of times, you will notice the rabbitmq memory usage increase even without queueing any actual messages.

**Comment by thedrow** [comment]
This does not look like a bug in Celery. However I'll need more information to determine that.
Can you reproduce this on another network? Without a VM?
This happens when there is not enough memory (or other resources) for the broker.

Also, please try to set the connect_timeout to a higher value, it should solve your issue.

**Comment by ananyashandilya1102** [comment]
@unclewizard @thedrow I am getting the same issue. Was this resolved? Can you help me with the solution?

**Comment by thedrow** [comment]
Without more details I can't be helpful.
I think this issue should be closed as it is not actionable.

If any of you think you have found the root cause of the issue please report a new issue about it.

**Comment by srinu230** [comment]
Hi i am also facing same issue

018-08-30 09:59:22.323 [warning] <0.21673.0> closing AMQP connection <0.21673.0> (127.0.0.1:53706 -> 127.0.0.1:5672, vhost: '/', user: 'guest'):
client unexpectedly closed TCP connection

i am using Brokerurl : amqp://guest:guest@localhost:5672//
and 
app = Celery('tasks', broker=broker_url(), include='task_helpers')
app.conf.update(
    CELERY_IGNORE_RESULT=True,
    CELERY_ENABLE_UTC=not settings.USE_LOCAL_TIMEZONE,
    CELERY_TIMEZONE='US/Eastern',
    CELERYD_PREFETCH_MULTIPLIER=BATCH_FLUSH_EVERY * 2,
    accept_content=['pickle', 'json', 'application/json'],
    CELERY_TASK_SERIALIZER='pickle',  # Default value; TODO: change this to msgpack
    CELERY_RESULT_SERIALIZER='pickle',  # Default value; TODO: change this to msgpack
	CELERY_ROUTES={
        'tasks.function1': {'queue': 'persistence'},
        'tasks.function2': {'queue': 'persistence'},
        'tasks.function3': {'queue': 'persistence'},
    },
    CELERY_DISABLE_RATE_LIMITS=True,  # We don't use them and it's recommended to disable if not using

    #BROKER_CONNECTION_TIMEOUT = 15
)

It is adding conneting to rabbitMq and not getting databack
when i run every time queues count increasing and same time getting above error.

client unexpectedly closed TCP connection
2018-08-30 09:59:22.255 [info] <0.21673.0> accepting AMQP connection <0.21673.0> (127.0.0.1:53706 -> 127.0.0.1:5672)
2018-08-30 09:59:22.257 [info] <0.21673.0> connection <0.21673.0> (127.0.0.1:53706 -> 127.0.0.1:5672): user 'guest' authenticated and granted access to vhost '/'
2018-08-30 09:59:22.323 [warning] <0.21673.0> closing AMQP connection <0.21673.0> (127.0.0.1:53706 -> 127.0.0.1:5672, vhost: '/', user: 'guest'):
client unexpectedly closed TCP connection

Could you Please helpme about why i am not getting databack
below list installed

alembic                      0.8.6    
amqp                         1.4.9    
anyjson                      0.3.3    
Babel                        2.6.0    
backports.ssl-match-hostname 3.5.0.1  
billiard                     3.3.0.23 
celery                       3.1.0    
certifi                      2018.8.24
cffi                         1.8.3    
chardet                      3.0.4    
commandr                     1.3.2    
crcmod                       1.7      
flower                       0.9.1    
GeoAlchemy2                  0.3.0    
idna                         2.7      
kombu                        3.0.37   
Mako                         1.0.7    
MarkupSafe                   1.0      
pip                          18.0     
psycopg2                     2.6.2    
pycparser                    2.18     
pydot                        1.2.4    
pymongo                      3.2.2    
pyparsing                    2.2.0    
python-editor                1.0.3    
pytz                         2018.5   
requests                     2.19.1   
setuptools                   40.2.0   
Shapely                      1.3.0    
SQLAlchemy                   1.0.12   
sqlalchemy-schemadisplay     1.3      
tornado                      4.2      
tzlocal                      1.2.2    
urllib3                      1.23     
wheel                        0.31.1

**Comment by xlanor** [comment]
Am getting the same issue.

> =INFO REPORT==== 16-Sep-2018::05:17:42 ===
> accepting AMQP connection <0.22297.5> (127.0.0.1:33752 -> 127.0.0.1:5672)
> 
> =INFO REPORT==== 16-Sep-2018::05:17:42 ===
> connection <0.22297.5> (127.0.0.1:33752 -> 127.0.0.1:5672): user 'guest' authenticated and granted access to vhost '/'
> 
> =WARNING REPORT==== 16-Sep-2018::05:17:42 ===
> closing AMQP connection <0.22297.5> (127.0.0.1:33752 -> 127.0.0.1:5672, vhost: '/', user: 'guest'):
> client unexpectedly closed TCP connection
> 

My logs are flooded with this, hundreds of times in the same time frame.

It always starts with this:

> **=ERROR REPORT==== 16-Sep-2018::05:49:26 ===
> closing AMQP connection <0.6725.20> (127.0.0.1:50504 -> 127.0.0.1:5672):
> missed heartbeats from client, timeout: 60s**
> 
> =INFO REPORT==== 16-Sep-2018::05:49:41 ===
> accepting AMQP connection <0.6824.20> (127.0.0.1:50842 -> 127.0.0.1:5672)
> 
> =INFO REPORT==== 16-Sep-2018::05:49:41 ===
> connection <0.6824.20> (127.0.0.1:50842 -> 127.0.0.1:5672): user 'guest' authenticated and granted access to vhost '/'
> 
> =WARNING REPORT==== 16-Sep-2018::05:49:41 ===
> closing AMQP connection <0.6824.20> (127.0.0.1:50842 -> 127.0.0.1:5672, vhost: '/', user: 'guest'):
> client unexpectedly closed TCP connection
> 
> =INFO REPORT==== 16-Sep-2018::05:49:41 ===
> accepting AMQP connection <0.6837.20> (127.0.0.1:50844 -> 127.0.0.1:5672)
> 
> =INFO REPORT==== 16-Sep-2018::05:49:41 ===
> connection <0.6837.20> (127.0.0.1:50844 -> 127.0.0.1:5672): user 'guest' authenticated and granted access to vhost '/'
> 
> =WARNING REPORT==== 16-Sep-2018::05:49:41 ===
> closing AMQP connection <0.6837.20> (127.0.0.1:50844 -> 127.0.0.1:5672, vhost: '/', user: 'guest'):
> client unexpectedly closed TCP connection


EDIT: Just saw #4895 . Will attempt to downgrade celery.

**Comment by auvipy** [comment]
are you using the latest celery?

**Comment by xlanor** [comment]
> are you using the latest celery?

I'm using 4.2.1 as reported in #4895 .  I will now attempt a downgrade to 4.1.1 to see if that resolves the issue.

Downgrading to 4.1.1 seems to have fixed the issue for me. I will test it further to see if the issue is still prevalent.

**Comment by auvipy** [comment]
then it might be a regression

**Comment by ghost** [comment]
Can reproduce exactly the same issue as xlanor and downgrading to 4.1.1 fixed it also for me.

**Comment by uniqueg** [comment]
So can I. Back to 4.1.1.

**Comment by ingmferrer** [comment]
I'm having this same issue and the downgrade to 4.1.1 seems to fix this.

**Comment by sloria** [comment]
We are also running into this issue. Downgrading to 4.1.1 fixes it, but this is not a feasible workaround for us because 4.2.0 contains some critical bugfixes, including a fix for #4223 .

Is the 4.1.x release line still supported? Will bug fixes be backported?

**Comment by thedrow** [comment]
We don't have the resources to maintain two versions currently.
We'd rather invest our effort in investigating bugs like these.

**Comment by sloria** [comment]
Understandable. Let us know if there is any further information we can provide to help diagnose this issue.

Thanks for all your hard work!

**Comment by thedrow** [comment]
At this point, any new information would be good.
Finding the offending commit would be awesome.

**Comment by sloria** [comment]
OK, I won't have time to do a code dive in the near future, unfortunately.

All I can say at the moment is that the issue occurs in our codebase when calling

```python
group(queue).apply_async()
```

When `queue` is large, a large payload (~500 Mb) payload is sent to RabbitMQ. The worker hangs, and eventually disconnects with the `client unexpectedly closed TCP connection` as detailed above.

**Comment by thedrow** [comment]
Oh, I think others have encountered that but I can't find the relevant issue.
How many tasks are we talking about?

**Comment by sloria** [comment]
@thedrow Last time I reproduced the issue, the group contained 18 total subtasks. One of the subtasks was a `chain` of 14 tasks.

**Comment by jshridha** [comment]
This issue occurs if my celery beat does not have any tasks that are scheduled to occur less than 1 minute apart. I currently have tasks scheduled to run every 5 minutes. After 60 seconds, the heartbeat is lost and the connection is dropped causing an infinite loop. I was able to track down the offending commit to https://github.com/celery/celery/commit/3af6a635cf90a4452db4e87c2326579ebad750c2.

Unfortunately it's a huge commit and I haven't been able to find where exactly the issue is. Hopefully that helps.

**Comment by thedrow** [comment]
The commit you are pointing to is a merge commit of master onto master.
The bug is obviously somewhere in there.

**Comment by ShaheedHaque** [comment]
We may be running into the same/similar issues and are running 4.2.1. A downgrade to 4.1.1 seemed to make the issue less frequent, but not completely disappear. We have a relatively simple set of usage cases:

1. We use Celery as a timer (i.e. the task is scheduled with a countdown and does a "resume" on the rest of our system, aka Django Viewflow using very little in the way of task logic, and minimal results to record etc.).
2. We use Celery to run expensive tasks (which record somewhat more information to the results backend).

We don't use group or chains etc.

Our results backend is Consul (so downgrading to 4.1.1. is problematic for us), so we can be sure that even use case 2 does not send excessive data to Consul as otherwise, Consul would complain. We have not been able to correlate the failure to anything specific we are asking of Celery; all we tend to see is that the celery log file stops recording the arrival of messages into master (and the error in the RabbitMQ log some time beforehand).

We also notice that typing a command into celery such as `celery status` often results in a failure, but that repeating the command seems to work as expected. (FWIW, my ignorant guess is that the incoming first failing command causes a retry internally which revives whatever event loop Celery uses. Or something like that).

**Comment by Sispheor** [comment]
Hello. Same issue here on v4.2. This will be fixed in v4.3?

**Comment by thedrow** [comment]
Unfortunately we haven't found the source of the problem yet.
If you can reproduce this, try to debug it.

**Comment by rob-deutsch** [comment]
For what its worth I encountered the exact same problem. Unfortunately I don't have the expertise/bandwidth to investigate at the moment, but I can confirm that my issue was fixed when I downgraded to 4.1.1.

**Comment by capital-G** [comment]
For me it still persists even when downgrading to 4.1.1

```
=WARNING REPORT==== 27-Jan-2020::14:44:02 ===
closing AMQP connection <0.5507.327> (10.42.7.61:34016 -> 10.42.6.226:5672, vhost: '/', user: 'rabbit'):
client unexpectedly closed TCP connection
=WARNING REPORT==== 27-Jan-2020::14:44:02 ===
closing AMQP connection <0.5335.327> (10.42.7.61:34014 -> 10.42.6.226:5672, vhost: '/', user: 'rabbit'):
client unexpectedly closed TCP connection
```

It turns out that it was triggered by my kubernetes live checks of each worker

Executing

```shell
celery -A my_project status
```

on a worker as suggested on https://github.com/celery/celery/issues/4079#issuecomment-437415370 triggers the warning above, so it seems something at the CLI status function is not closing properly

**Comment by djsuszi** [comment]
anyone test it with 4.4 ? is it working ? or still version > 4.1.1 will have this issue?

**Comment by switchwebmaster** [comment]
Version 4.4.0 (cliffs) here and constantly seeing the connections closed logs on RabbitMQ and also in Celery. From my case I have tasks that could take 1h to execute... And Celery constantly dies and process the same tasks repeatedly. There are some possible configurations fault in Celery?

**Comment by waleedsamy** [comment]
+1

**Comment by gitttt** [comment]
+1

**Comment by auvipy** [comment]
please feel free to contribute

**Comment by anandvlps** [comment]
Getting the same issue after recent updates on my mac (it was working fine previously).
**Mac version** - Mac Catelina 10.15.6
**Rabbitmq Details:**
RabbitMQ version: 3.8.8
Node name: rabbit@localhost
Erlang configuration: Erlang/OTP 23 [erts-11.0.4] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:128] [hipe] [dtrace]

**Celery Details:**
amqp==2.6.1
billiard==3.6.3.0
celery==4.4.7
kombu==4.6.11
pika==1.1.0
pytz==2020.1
redis==3.5.3
vine==1.3.0

Sample Program used - tasks.py:

```
from celery import Celery
from celery import platforms

#platforms.C_FORCE_ROOT = True

app = Celery('tasks', broker='pyamqp://guest@localhost//')
#app = Celery('tasks', broker='redis://localhost:6379/0')

@app.task
def add(x, y):
    return x + y
```

On running the command - **celery -A tasks worker --loglevel=info**

Getting the following output -

 ```
-------------- celery@PUN-IT000604.local v4.4.7 (cliffs)
--- ***** ----- 
-- ******* ---- macOS-10.15.6-x86_64-i386-64bit 2020-09-20 01:05:55
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         tasks:0x7f9aac182af0
- ** ---------- .> transport:   amqp://guest:**@localhost:5672//
- ** ---------- .> results:     disabled://
- *** --- * --- .> concurrency: 8 (prefork)
-- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
--- ***** ----- 
 -------------- [queues]
                .> celery           exchange=celery(direct) key=celery
                

[tasks]
  . tasks.add

[2020-09-20 01:06:00,033: ERROR/MainProcess] consumer: Cannot connect to amqp://guest:**@127.0.0.1:5672//: timed out.
Trying again in 2.00 seconds... (1/100)

[2020-09-20 01:06:06,055: ERROR/MainProcess] consumer: Cannot connect to amqp://guest:**@127.0.0.1:5672//: timed out.
Trying again in 4.00 seconds... (2/100)

[2020-09-20 01:06:14,082: ERROR/MainProcess] consumer: Cannot connect to amqp://guest:**@127.0.0.1:5672//: timed out.
Trying again in 6.00 seconds... (3/100)
```

Logs from /usr/local/var/log/rabbitmq/rabbit@localhost.log 
```
2020-09-20 01:14:51.216 [info] <0.1733.0> accepting AMQP connection <0.1733.0> (127.0.0.1:64284 -> 127.0.0.1:5672)
2020-09-20 01:14:56.218 [warning] <0.1733.0> closing AMQP connection <0.1733.0> (127.0.0.1:64284 -> 127.0.0.1:5672):
client unexpectedly closed TCP connection
2020-09-20 01:14:57.238 [info] <0.1743.0> accepting AMQP connection <0.1743.0> (127.0.0.1:64285 -> 127.0.0.1:5672)
2020-09-20 01:15:02.244 [warning] <0.1743.0> closing AMQP connection <0.1743.0> (127.0.0.1:64285 -> 127.0.0.1:5672):
client unexpectedly closed TCP connection
2020-09-20 01:15:05.255 [info] <0.1754.0> accepting AMQP connection <0.1754.0> (127.0.0.1:64290 -> 127.0.0.1:5672)
2020-09-20 01:15:10.262 [warning] <0.1754.0> closing AMQP connection <0.1754.0> (127.0.0.1:64290 -> 127.0.0.1:5672):
client unexpectedly closed TCP connection
```

But I tested that rabbitmq is working fine, i am able to send and receive messages using the following example -

send.py -
```
#!/usr/bin/env python
import pika

connection = pika.BlockingConnection(
    pika.ConnectionParameters(host='localhost'))
channel = connection.channel()

channel.queue_declare(queue='hello')

channel.basic_publish(exchange='', routing_key='hello', body='Hello World!')
print(" [x] Sent 'Hello World!'")
connection.close()
```
receive.py -

```
#!/usr/bin/env python
import pika, sys, os

def main():
    connection = pika.BlockingConnection(pika.ConnectionParameters(host='localhost'))
    channel = connection.channel()

    channel.queue_declare(queue='hello')

    def callback(ch, method, properties, body):
        print(" [x] Received %r" % body)

    channel.basic_consume(queue='hello', on_message_callback=callback, auto_ack=True)

    print(' [*] Waiting for messages. To exit press CTRL+C')
    channel.start_consuming()

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('Interrupted')
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)
```

Output of send.py - 
```
python send.py 
 [x] Sent 'Hello World!'
```

Output of receive.py -
```
python receive.py 
 [*] Waiting for messages. To exit press CTRL+C
 [x] Received b'Hello World!'
```

Please help me to debug this.

**Comment by auvipy** [comment]
why you are using pika with celery? celery has py-amqp

**Comment by anandvlps** [comment]
Yes celery has py-amqp, but it is not able to make the connection.
i wanted to debug where the issue is whether it is from celery side or from rabbitmq side. As celery is not able to make the connection, i wanted to check with other client which i found as pika.

**Comment by ba1dr** [comment]
For me I had the same issue with connection dropped after 25  minutes with the long running task. This looked similar for both parties - RabbitMQ and celery reported that another peer reset the connection.
Finally I've discovered that this was related to Docker (I run my celery inside Docker). Quick workaround was to run containers with `--net host` and it seems to work now.

For additional info see your `net.ipv4.tcp_keepalive_time` setting (sysctl) and the following article: https://success.mirantis.com/article/ipvs-connection-timeout-issue

**Reply by ooyamatakehisa** [reply]
Were you using `docker-compose` and was the connection between containers inside `docker-compose`? or rabbitmq wasn't inside a container?

**Reply by ba1dr** [reply]
No, I did not use `docker-compose` for that project. The application was in one container and RabbitMQ was on a separate server (not in docker).

**Reply by ooyamatakehisa** [reply]
Thank you for the quick reply!
I'm using docker-compose so that's not the case for me😭

**Comment by auvipy** [comment]
> For me I had the same issue with connection dropped after 25 minutes with the long running task. This looked similar for both parties - RabbitMQ and celery reported that another peer reset the connection.
> Finally I've discovered that this was related to Docker (I run my celery inside Docker). Quick workaround was to run containers with `--net host` and it seems to work now.
> 
> For additional info see your `net.ipv4.tcp_keepalive_time` setting (sysctl) and the following article: https://success.mirantis.com/article/ipvs-connection-timeout-issue

based on the fact should we close the issue or you suggest some improvement in celery?

**Comment by auvipy** [comment]
> For me I had the same issue with connection dropped after 25 minutes with the long running task. This looked similar for both parties - RabbitMQ and celery reported that another peer reset the connection.
> Finally I've discovered that this was related to Docker (I run my celery inside Docker). Quick workaround was to run containers with `--net host` and it seems to work now.
> 
> For additional info see your `net.ipv4.tcp_keepalive_time` setting (sysctl) and the following article: https://success.mirantis.com/article/ipvs-connection-timeout-issue

or we should mention this in the docs?

**Comment by boldrack** [comment]
Unfortunately This issue still persists in version 5.0.5 . Is there any workaround for this  ?

**Comment by auvipy** [comment]
> Unfortunately This issue still persists in version 5.0.5 . Is there any workaround for this ?

did you check https://github.com/celery/celery/issues/4355#issuecomment-702626238 ?

**Comment by ShanthaDodmane** [comment]
The link mentioned in this comment https://github.com/celery/celery/issues/4355#issuecomment-702626238  is asking us to login to salesforce.com :(

**Comment by nicolaerosia** [comment]
I'm seeing this when I do celery ping, e.g.
```
celery inspect ping -d celery@${HOSTNAME}
```

RabbitMQ:
```
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.299663+00:00 [info] <0.13830.5> accepting AMQP connection <0.13830.5> (10.10.186.247:36212 -> 10.10.196.154:5672)
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.305745+00:00 [info] <0.13830.5> connection <0.13830.5> (10.10.186.247:36212 -> 10.10.196.154:5672): user 'root' authenticated and granted access to vhost '/'
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.345957+00:00 [info] <0.13847.5> accepting AMQP connection <0.13847.5> (10.10.186.247:36214 -> 10.10.196.154:5672)
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.365854+00:00 [info] <0.13847.5> connection <0.13847.5> (10.10.186.247:36214 -> 10.10.196.154:5672): user 'root' authenticated and granted access to vhost '/'
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.414558+00:00 [warn] <0.13847.5> closing AMQP connection <0.13847.5> (10.10.186.247:36214 -> 10.10.196.154:5672, vhost: '/', user: 'root'):
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.414558+00:00 [warn] <0.13847.5> client unexpectedly closed TCP connection
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.414734+00:00 [warn] <0.13830.5> closing AMQP connection <0.13830.5> (10.10.186.247:36212 -> 10.10.196.154:5672, vhost: '/', user: 'root'):
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.414734+00:00 [warn] <0.13830.5> client unexpectedly closed TCP connection
```

I will try to debug celery, it feels like the connection is not properly closed after being opened.

**Comment by nicolaerosia** [comment]
OK, I did debug a bit and found the culprit to be the connection pool in kombu:
```
broker_pool_limit = 0
```

results in a nice RabbitMQ log:
```
[pod/rabbitmq-0/rabbitmq] 2021-09-20 10:47:52.021612+00:00 [info] <0.22524.5> accepting AMQP connection <0.22524.5> (127.0.0.1:57356 -> 127.0.0.1:5672)
[pod/rabbitmq-0/rabbitmq] 2021-09-20 10:47:54.134131+00:00 [info] <0.22524.5> connection <0.22524.5> (127.0.0.1:57356 -> 127.0.0.1:5672): user 'root' authenticated and granted access to vhost '/'
[pod/rabbitmq-0/rabbitmq] 2021-09-20 10:48:01.602229+00:00 [info] <0.22546.5> accepting AMQP connection <0.22546.5> (127.0.0.1:57458 -> 127.0.0.1:5672)
[pod/rabbitmq-0/rabbitmq] 2021-09-20 10:48:03.962610+00:00 [info] <0.22546.5> connection <0.22546.5> (127.0.0.1:57458 -> 127.0.0.1:5672): user 'root' authenticated and granted access to vhost '/'
[pod/rabbitmq-0/rabbitmq] 2021-09-20 10:48:08.972621+00:00 [info] <0.22546.5> closing AMQP connection <0.22546.5> (127.0.0.1:57458 -> 127.0.0.1:5672, vhost: '/', user: 'root')
[pod/rabbitmq-0/rabbitmq] 2021-09-20 10:48:17.274228+00:00 [info] <0.22524.5> closing AMQP connection <0.22524.5> (127.0.0.1:57356 -> 127.0.0.1:5672, vhost: '/', user: 'root')
```

I guess the problem lies in the teardown of the pool? @auvipy does this ring any bell? :)

anyway, I don't understand why for a ping we need two connections, any idea?

**Comment by auvipy** [comment]
> OK, I did debug a bit and found the culprit to be the connection pool in kombu:
> 
> ```
> broker_pool_limit = 0
> ```

does changing the value here improve anything?

**Comment by nicolaerosia** [comment]
yes, the connections are closed gracefully, I no longer see the warnings in RabbitMQ log:
```
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.414558+00:00 [warn] <0.13847.5> closing AMQP connection <0.13847.5> (10.10.186.247:36214 -> 10.10.196.154:5672, vhost: '/', user: 'root'):
[pod/rabbitmq-0/rabbitmq] 2021-09-20 07:18:21.414558+00:00 [warn] <0.13847.5> client unexpectedly closed TCP connection
```

**Comment by auvipy** [comment]
instead of closing I am moving this to discussion

**Comment by iamnotagentleman** [comment]
@auvipy any progress for this problem? i think this problem cause high cpu usages too.

**Comment by kernelshard** [comment]
Hello all, any solution? Getting same for the celery 4.4.2

**Reply by invi-code** [reply]
hello everyone did anyone got the solution?

**Comment by unitysipu** [comment]
iirc broker_pool_limit = 0 fixed this for me as well

**Comment by iglesiasroberto14** [comment]
Same here. Happening at least since version 4.1, when we decided to disable the broker pool as a workaround (`broker_pool_limit=0`). Now we upgraded to the latest available version (5.2.7), but it's still happening. I wasn't able to find the root cause. We need to enable broker pool, because we have too much tcp overhead, every task starts and finishes a new TCP connection, resulting in tens of thousands of connections every day.

Versions:
```
/ # pip freeze | grep -E '(amqp|celery|kombu)'
amqp==5.1.1
celery==5.2.7
kombu==5.2.4
```

I'd be glad to collaborate, but celery and kombu projects are a bit messy, and I can't find where the problem is. So... any hint is welcomed.

**Reply by ppiekars** [reply]
I did some steps as a proof of concept:

I had hard time finding any logic that would close the producer pool explicitly. I've put `kombu.pools.reset()` inside `on_close` method  for the WorkController class. However, it looked like for some cases it was closing the underlying TCP connection directly, without sending the appropriate close message first. As a quick fix, I've requested explicit resource closure by putting `res.close()` inside `force_close_all` method, just before `self.collect_resource(res)` (in kombu library). This did the trick and now starting up and shutting down workers does not result in unexpectedly closing TCP connections.
There's one catch. There is still no logic on closing connections within forked worker processes. I've put `kombu.pools.reset()` inside billiard pool `receive` function, just before `raise SystemExit(EX_FAILURE)`. 
Now, forked processes close the connections gracefully. 

These changes cannot be implemented as is for obvious reasons 😄  But maybe it will help find the correct path for solving the issue.

@auvipy maybe you could chime in on that, commenting whether it makes sense to implement this kind of pool closing during the shutdown, both in the main process and forked processes as well. Same thing can be implemented as a global pool cleanup during various commands like `status` or `inspect ping`.

**Reply by auvipy** [reply]
It would be easier for me to comment/share my opinion if you can share some working/draft code example. of course with relevant test in the form of a PR?

**Comment by ooyamatakehisa** [comment]
I have the same error.
```
2023-07-11 04:19:03.601367+00:00 [warning] <0.986.0> closing AMQP connection <0.986.0> (172.18.0.5:35116 -> 172.18.0.2:5672, vhost: '/', user: 'guest'):
2023-07-11 04:19:03.601367+00:00 [warning] <0.986.0> client unexpectedly closed TCP connection
```
I'm not sure sharing my environmet helps but the following is my env just in case.
- using docker-compose
- broker is `rabbitmq:3.12.0-management` container
- celery version is `celery==5.3.1`
- worker's base image is `centos/python-38-centos7`

**Comment by MugdhaHardikar-GSLab** [comment]
I am also facing similar issue. My detailed analysis is mentioned [here](https://github.com/celery/celery/discussions/8594)
Versions:
rabbitmq: 3.12.7
Celery : 5.1.2
Mongo: 5.0.21

Is it recommended to change the celery version?
I have tried setting "broker_pool_limit" during worker creation. It has not helped me.

**Reply by MugdhaHardikar-GSLab** [reply]
Adding below in celery app configuration solved my issue:

+            broker_pool_limit=0,
+            broker_transport_options={'confirm_publish': True},

**Reply by bilgehan-biricik** [reply]
That solution worked for me! Thank you!

**Comment by aditya-shrivastavv** [comment]
Hey, I get same logs in rabbitmq:

```
2024-08-14 07:29:38.532754+00:00 [warning] <0.7891.0> closing AMQP connection <0.7891.0> (10.244.0.30:58308 -> 10.244.0.5:5672, vhost: '/', user: 'guest'):             │
│ 2024-08-14 07:29:38.532754+00:00 [warning] <0.7891.0> client unexpectedly closed TCP connection 
```

I am trying to rabbitmq like:

```py
def main():
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host='rabbitmq')
    )
    channel = connection.channel()

    def callback(chan, method, props, body):
        err = email.notify(body)
        if err:
            chan.basic_nack(delivery_tag=method.delivery_tag)
        else:
            chan.basic_ack(delivery_tag=method.delivery_tag)

    channel.basic_consume(
        queue=os.environ.get("MP3_QUEUE"),
        on_message_callback=callback
    )

    print("Waiting for messages. To leave: CTRL+C")
    channel.start_consuming()


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("Exiting...")
        # Graceful exit
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)
```

can anyone help where could be the problem be? At line `channel.start_consuming()` the logs say:

```
Connection unexpectedly closed
Connection unexpectedly closed
Connection unexpectedly closed
Connection unexpectedly closed
...
```

**Comment by amn-max** [comment]
Any update on this?

am facing the same issue

```
2024-09-11 13:22:43.165224+00:00 [info] <0.1137.0> Waiting for Mnesia tables for 30000 ms, 9 retries left
2024-09-11 13:22:43.165385+00:00 [info] <0.1137.0> Successfully synced tables from a peer
2024-09-11 13:23:05.424995+00:00 [info] <0.1152.0> accepting AMQP connection <0.1152.0> (172.21.0.8:40156 -> 172.21.0.10:5672)
2024-09-11 13:23:05.427847+00:00 [info] <0.1152.0> connection <0.1152.0> (172.21.0.8:40156 -> 172.21.0.10:5672): user 'guest' authenticated and granted access to vhost 'main-vhost'
2024-09-11 13:23:05.440209+00:00 [info] <0.1169.0> accepting AMQP connection <0.1169.0> (172.21.0.8:40170 -> 172.21.0.10:5672)
2024-09-11 13:23:05.443586+00:00 [info] <0.1169.0> connection <0.1169.0> (172.21.0.8:40170 -> 172.21.0.10:5672): user 'guest' authenticated and granted access to vhost 'main-vhost'
2024-09-11 13:23:06.289555+00:00 [info] <0.1181.0> accepting AMQP connection <0.1181.0> (172.21.0.7:57302 -> 172.21.0.10:5672)
2024-09-11 13:23:06.291799+00:00 [info] <0.1181.0> connection <0.1181.0> (172.21.0.7:57302 -> 172.21.0.10:5672): user 'guest' authenticated and granted access to vhost 'main-vhost'
2024-09-11 13:23:06.304370+00:00 [info] <0.1198.0> accepting AMQP connection <0.1198.0> (172.21.0.7:57312 -> 172.21.0.10:5672)
2024-09-11 13:23:06.307053+00:00 [info] <0.1198.0> connection <0.1198.0> (172.21.0.7:57312 -> 172.21.0.10:5672): user 'guest' authenticated and granted access to vhost 'main-vhost'
2024-09-11 13:23:06.529468+00:00 [warning] <0.1169.0> closing AMQP connection <0.1169.0> (172.21.0.8:40170 -> 172.21.0.10:5672, vhost: 'main-vhost', user: 'guest'):
2024-09-11 13:23:06.529468+00:00 [warning] <0.1169.0> client unexpectedly closed TCP connection
2024-09-11 13:23:06.529725+00:00 [warning] <0.1152.0> closing AMQP connection <0.1152.0> (172.21.0.8:40156 -> 172.21.0.10:5672, vhost: 'main-vhost', user: 'guest'):
2024-09-11 13:23:06.529725+00:00 [warning] <0.1152.0> client unexpectedly closed TCP connection
2024-09-11 13:23:07.386494+00:00 [warning] <0.1181.0> closing AMQP connection <0.1181.0> (172.21.0.7:57302 -> 172.21.0.10:5672, vhost: 'main-vhost', user: 'guest'):
2024-09-11 13:23:07.386494+00:00 [warning] <0.1181.0> client unexpectedly closed TCP connection
2024-09-11 13:23:07.386494+00:00 [warning] <0.1198.0> closing AMQP connection <0.1198.0> (172.21.0.7:57312 -> 172.21.0.10:5672, vhost: 'main-vhost', user: 'guest'):
2024-09-11 13:23:07.386494+00:00 [warning] <0.1198.0> client unexpectedly closed TCP connection
```


Versions:
rabbitmq: 3.13.7 - broker
Celery : 5.4.0
redis: 5.0.6 FROM bitnami/redis:7.0.15 as result backend

for me `broker_pool_limit=0` and `broker_transport_options={'confirm_publish': True}` did not work

Am running all the services in docker, all tasks seems to be executing fine no issues there

**Reply by guoard** [reply]
> for me `broker_pool_limit=0` and `broker_transport_options={'confirm_publish': True}` did not work

Same issue with version `5.5.3`.

**Reply by auvipy** [reply]
please try with celery v5.6.2 and report again

**Reply by lchojnacki** [reply, marked as answer]
I'm on Celery 5.6.2, and RabbitMQ still warns about "client unexpectedly closed TCP connection." This is the command I'm running:

```
celery -A project_name.celery inspect ping
```

RabbitMQ docker image: `rabbitmq:4.1.0-management`.

EDIT:

After applying the above options (`broker_pool_limit=0`, `broker_transport_options={"confirm_publish": True}`), the warning is gone.
